### ISSUE TITLE
B300: fuse vocab-parallel CE label-smoothing statistics into the forward kernel

### ISSUE THREAD

**Issue by heiheiha798** [heiheiha798/Liger-Kernel#16]
## Problem

`LigerVocabParallelCrossEntropy` and `LigerMegatronCrossEntropy` already compute the global logit maximum and visit every local-vocabulary logit in the Triton forward kernel. When `label_smoothing > 0`, the wrapper nevertheless materializes another full FP32 expression:

```python
log_probs_local = torch.log(exp_buf) - torch.log(sum_exp_local).unsqueeze(-1)
```

It then reduces that `[BT, V_local]` tensor to one statistic per row. This adds per-element logarithms, full-buffer reads/writes, allocator traffic, and eager launches after the kernel has already loaded the required logits.

The memory scale is material. At `BT=8192` and `V_local=131072`, one FP32 `[BT, V_local]` tensor is 4 GiB. At TP=8 with global vocabulary 131072, it is still 512 MiB per rank. These are source-derived sizes, not measured B300 performance claims.

## Proposed SM103 path

For exact Blackwell Ultra (`infer_device_arch(actual_device_id) == "blackwell_ultra"`) and `label_smoothing > 0`, fuse the sufficient statistic into `liger_vocab_parallel_ce_forward_kernel`:

```text
sum_shifted_local = sum_i (x_i - global_max)
sum_i log_softmax(x_i) = sum_shifted - K * log(sum_exp_global)
```

The kernel should emit only one FP32 scalar per row in addition to its existing outputs.

- `global` smoothing: AllReduce the `[BT]` shifted-logit sum and use `K=V_global`. This replaces the existing AllReduce of `[BT] sum_log_probs`; collective count and byte volume must not increase.
- `partition` smoothing: use the local shifted-logit sum and `K=V_local`; no collective is added.
- Resolve the architecture from the actual logits CUDA device, including non-default device IDs, and launch the eligible kernel under that device's context.
- Check `label_smoothing > 0` before architecture dispatch so the zero-smoothing public path pays no new host lookup. Compile the accumulator and store away for zero smoothing.
- Preserve the current eager `torch.log(exp_buf)` path for non-SM103 devices and unsupported cases.

This is an algebraic sufficient-statistic fusion, not a change to the saved FP32 `exp_buf`, backward formula, label-smoothing formulas/modes, collective topology, block/warp policy, public API, or other backends.

## Scope

Tracked files are limited to:

- `src/liger_kernel/ops/vocab_parallel_cross_entropy.py`
- `test/megatron/test_cross_entropy.py`

Use an ignored temporary exact-source harness for profiling and performance evidence. Do not modify the shared benchmark script or committed benchmark data, change exp-buffer precision, tune the producer schedule, add a new collective, touch fused-linear CE, or generalize this to other CE implementations.

Open PR #1331 overlaps only through launch-device context and may require a mechanical rebase. Open PR #1371 implements fused-linear vocab-parallel CE in separate files and is not this optimization.

## Correctness

Use independent formulas as the primary oracle, not only source-exact legacy Liger.

- TP=1: compare both public classes against FP32 `torch.nn.functional.cross_entropy`, including output and logits gradients from independent non-unit upstream gradients.
- Cover PyTorch and Megatron smoothing formulas, `label_smoothing={1e-4,0.1,0.5,0.9}`, FP16/BF16/FP32, `V={123,4096,32000,65536,131072}`, non-power-of-two `BT`, mixed/all ignored rows, and finite logits with large negative gaps.
- TP>=2 global mode: compare output and each rank's gradient slice with a full-vocabulary PyTorch oracle.
- TP>=2 partition mode: add an explicit independent oracle using global `logsumexp`, each rank's local `sum(x-global_max)`, local `K`, the documented alpha rescaling, and the existing target adjustment. The current test skips forward and gradient parity for this branch, so source-exact self-comparison is insufficient.
- Require at least TP=2 multi-GPU correctness for both modes before treating the result as upstream-ready; TP=4/8 are additional coverage when an established allocation is available.
- Verify `label_smoothing=0` remains unchanged, collective count/volume stays unchanged, exact SM103 dispatch uses the input's nonzero device ID, and non-SM103/non-CUDA paths remain legacy.
- Run the complete Megatron CE and monkey-patch tests plus `make checkstyle`. Do not relax existing tolerances to hide reduction-order differences.

The algebraic path can remain finite where the current `log(exp(x-max))` underflows. Treat that only as oracle-backed correctness evidence; do not claim or broaden into a separate numerical-semantics change.

## B300 validation

Run all GPU work through Slurm. Use clean detached baseline/candidate worktrees in fresh processes within the same B300 allocation, with exact import paths, immutable logits/targets/upstream gradients, fixed collective-safe iteration counts, sufficient compilation warmup, and at least seven alternating paired rounds. Record every round plus p20/median/p80, exact commits, job/node/GPU, driver, CUDA, PyTorch, Triton, NCCL, commands, environment, and raw outputs.

Profile first at BF16 `BT=8192, V_local=131072`, TP=1, smoothing 0.1:

- attribute the current log/subtract/reduction launches and full-buffer traffic;
- compare the modified Triton kernel's registers, occupancy, spills, DRAM/L2 traffic, and duration;
- verify that the additional row reduction does not create a producer-kernel regression.

Primary TP=1 matrix:

- `BT={1024,4096,8192}`
- `V={32768,65536,131072}`
- BF16, smoothing 0.1
- public forward, backward control, directly measured full, peak allocated/reserved memory, and launch count

Controls include smoothing `{0,0.5}`, FP16/FP32, `V=32000`, ignore-heavy rows, and numerical-stress inputs. Run the same exact-source method at TP=2 for both global and partition smoothing; TP=4/8 are optional only when an established multi-GPU B300 allocation is available. Verify NCCL collective count and bytes rather than inferring them from source.

## Acceptance gate

Proceed only if baseline profiling confirms the post-kernel full-buffer chain is material and the fused statistic passes the independent oracle.

- At least 6/9 TP=1 BF16 primary forward medians improve by at least 20%, including both predeclared long-token `V=131072` points.
- At least 5/9 corresponding direct-full medians improve by at least 10%.
- TP=2 must show a stable public-path improvement in both global and partition modes.
- Peak allocated memory must fall by at least one `[BT,V_local]` FP32 buffer on every primary width where that intermediate is material; report absolute MiB and ratios.
- No primary/control/fallback point may regress by more than 5%. Zero-smoothing, backward-only, collective count/volume, and non-target paths may not regress by more than 3%.
- No new workspace-sized allocation or per-shape schedule is allowed.

If the added reduction spills, changes smoothing support/tolerances, fails distributed correctness, or does not improve the public path, record a no-go without broadening scope or weakening the gates.

## Why this is B300-relevant

This removes GiB-scale FP32 intermediate traffic and a full-vocabulary SFU/reduction chain from a public distributed-training loss, while reusing data already live in the forward Triton pass and preserving communication complexity. The expected gain is a profile-first hypothesis to validate on B300/SM103, not an established benchmark result.

**Comment by heiheiha798** [comment]
## Cycle 13 result: aborted because TP2 validation is unavailable

The single-GPU B300 profile-first gate was positive, but this cycle is **not** being presented as a mergeable implementation. The immutable acceptance plan requires independent TP=2 global and partition correctness before a candidate commit, and the active Slurm QoS rejects every two-GPU allocation even when the user has no running or pending jobs.

### Positive single-GPU profile gate

- Slurm job: `5069`
- Node / GPU: `dev-slurm`, NVIDIA B300 SXM6 AC, compute capability 10.3, 148 SMs
- Software: driver 580.126.09, CUDA 13.0, PyTorch 2.13.0+cu130, Triton 3.7.1, NCCL 2.29.7, Nsight Compute 2025.3.1
- Repository state: branch `b300/vpce-sm103-label-smoothing-stat-fusion` at baseline commit `124fb8a2a443c1118f7caeff0203aea78424eb2b`; the candidate was an ignored temporary harness used before any tracked commit
- Profile point: BF16, TP=1, `BT=8192`, `V=131072`, smoothing 0.1
- Raw output: `.cache/b300-vpce-smoothing/gate-5069.out`
- Raw NCU: `.cache/b300-vpce-smoothing/gate-ncu-baseline-5069.csv` and `.cache/b300-vpce-smoothing/gate-ncu-candidate-5069.csv`

The baseline timeline measured the eager smoothing chain (`log`, subtract, and reduction) after the Triton producer and showed GiB-scale temporary allocation. The temporary algebraic statistic fusion passed its focused FP16/BF16/FP32 checks for both smoothing formulas and all four declared smoothing values. Across seven alternating rounds, paired median ratios were:

| Operation | Baseline / temporary candidate |
|---|---:|
| Forward | 3.094x |
| Backward | 1.000x |
| Direct full | 2.022x |

At the profile point, incremental peak allocated memory fell from 12,885,065,728 bytes to 4,295,262,208 bytes for forward, eliminating the full-buffer smoothing intermediates. Direct-full incremental peak allocation fell from 12,885,065,728 bytes to 6,442,516,480 bytes.

### Blocking TP2 gate

The prepared command was:

```bash
sbatch .cache/b300-vpce-smoothing/precommit_tests.sbatch
```

It requested two GPUs and retained the complete Megatron CE and monkey-patch suites, including the new independent TP=2 global/partition oracle and collective-count checks. Slurm rejected the submission twice before creating any job:

```text
sbatch: error: QOSMaxGRESPerUser
sbatch: error: Batch job submission failed: Job violates accounting/QOS policy (job submit limit, user's size and/or time limits)
```

The first rejection occurred while unrelated one-GPU work occupied the user-level queue. After those allocations completed, the one approved deferred submission received the same rejection; an immediate read-only `squeue -u $USER` check then showed zero running or pending jobs. This establishes a one-GPU QoS ceiling rather than temporary capacity contention. No TP2 job ID exists.

### Disposition

Cycle 13 is `PR_ABORTED` for environment incapability, not a performance no-go. The mandatory distributed correctness gate was not reduced or bypassed. There is no implementation commit, push, or draft PR; the fork branch remains at baseline `124fb8a2a443c1118f7caeff0203aea78424eb2b`. The bounded source/test experiment remains unpublished and no upstream write was made.
